pkg/cloud/services/provider/provider.go

```diff
@@ -109,9 +109,22 @@ func NewClient(cloud clientconfig.Cloud, caCert []byte) (*gophercloud.ProviderCl
 	}
 	err = openstack.Authenticate(provider, *opts)
 	if err != nil {
-		return nil, nil, fmt.Errorf("providerClient authentication err: %v", err)
+		return nil, nil, "", fmt.Errorf("providerClient authentication err: %v", err)
 	}
-	return provider, clientOpts, nil
+
+	// Set ProjectID for consumers of the ProviderClient
+	// This is only needed when application credentials were used, because they
+	// do not specify a `project_id` in the clientconfig.Cloud.
```

Comment:
As noted by @jichenjc this is how we always get project id, not just for application credentials.